File: csharp/lib/Property/ImuDataUnitProperty.cs
/*
 * SPDX-FileCopyrightText: 2017-2023 Sony Semiconductor Solutions Corporation
 *
 * SPDX-License-Identifier: Apache-2.0
 */

using MessagePack;

namespace SensCord
{
    /// <summary>
    /// Property for obtaining unit of RawData.
    /// </summary>
    [MessagePackObject]
    public class ImuDataUnitProperty : IBaseProperty
    {
        /// <summary>
        /// Property key.
        /// </summary>
        public static readonly string ConstKey = "imu_data_unit_property";

        /// <summary>
        /// Property key.
        /// </summary>
        [IgnoreMember]
        public string Key { get; } = ConstKey;

        /// <summary>
        /// Unit of data of acceleration.
        /// </summary>
        [Key("acceleration")]
        public AccelerationUnit Acceleration { get; set; }

        /// <summary>
        /// Unit of data of angular velocity.
        /// </summary>
        [Key("angular_velocity")]
        public AngularVelocityUnit AngularVelocity { get; set; }

        /// <summary>
        /// Unit of data of the magnetic field.
        /// </summary>
        [Key("magnetic_field")]
        public MagneticFieldUnit MagneticField { get; set; }

        /// <summary>
        /// Unit of data of the orientation.
        /// </summary>
        [Key("orientation")]
        public OrientationUnit Orientation { get; set; }

        /// <summary>
        /// To string.
        /// </summary>
        /// <returns>A string that represents the current object.</returns>
        public override string ToString()
        {
            return string.Format(
                "{0}(accel={1}, angular={2}, magnetic={3}, orientation={4})",
                this.Key, this.Acceleration, this.AngularVelocity, this.MagneticField, this.Orientation);
        }
    }

    /// <summary>
    /// Units used for acceleration.
    /// </summary>
    public enum AccelerationUnit
    {
        /// <summary>
        /// sensor not supported.
        /// </summary>
        NotSupported,

        /// <summary>
        /// Unit:[G].
        /// </summary>
        Gravitational,

        /// <summary>
        /// Unit:[m/s2].
        /// </summary>
        MetrePerSecondSquared,
    }

    /// <summary>
    /// Units used for angular velocity.
    /// </summary>
    public enum AngularVelocityUnit
    {
        /// <summary>
        /// sensor not supported.
        /// </summary>
        NotSupported,

        /// <summary>
        /// Unit:[deg/s].
        /// </summary>
        DegreePerSecond,

        /// <summary>
        /// Unit:[rad/s].
        /// </summary>
        RadianPerSecond,
    }

    /// <summary>
    /// Units used for magnetic field.
    /// </summary>
    public enum MagneticFieldUnit
    {
        /// <summary>
        /// sensor not supported.
        /// </summary>
        NotSupported,

        /// <summary>
        /// Unit:[gauss].
        /// </summary>
        Gauss,

        /// <summary>
        /// Unit:[uT].
        /// </summary>
        MicroTesla,
    }

    /// <summary>
    /// Units used for orientation.
    /// </summary>
    public enum OrientationUnit
    {
        /// <summary>
        /// sensor not supported.
        /// </summary>
        NotSupported,

        /// <summary>
        /// Unit:[deg].
        /// </summary>
        Degree,

        /// <summary>
        /// Unit:[rad].
        /// </summary>
        Radian,
    }
}

File: java/senscord/src/senscord_properties/ImuDataUnitProperty.java
/*
 * SPDX-FileCopyrightText: 2019 Sony Semiconductor Solutions Corporation
 *
 * SPDX-License-Identifier: Apache-2.0
 */

package senscord_properties;

import com.sun.jna.Structure;
import java.util.Arrays;
import java.util.List;

/**
 * Property for obtaining unit of RawData.
 * 
 * @see "senscord::ImuDataUnitProperty"
 */
public class ImuDataUnitProperty extends Structure {
    public int acceleration;
    public int angular_velocity;
    public int magnetic_field;
    public int orientation;

    public static class ByReference extends ImuDataUnitProperty implements Structure.ByReference {
    }

    @Override
    protected List<String> getFieldOrder() {
        return Arrays.asList("acceleration", "angular_velocity", "magnetic_field", "orientation");
    }

    /**
     * Units used for acceleration.
     * 
     * @see "senscord::AccelerationUnit"
     */
    public static interface AccelerationUnit {
        /**
         * Not supported.
         */
        public static final int SENSCORD_ACCELERATION_NOT_SUPPORTED = 0;

        /**
         * Unit : [G].
         */
        public static final int SENSCORD_GRAVITATIONAL = 1;

        /**
         * Unit : [m/s^2].
         */
        public static final int SENSCORD_METER_PER_SECOND_SQUARED = 2;
    }

    /**
     * Units used for angular velocity.
     * 
     * @see "senscord::AngularVelocityUnit"
     */
    public static interface AngularVelocityUnit {
        /**
         * Not supported.
         */
        public static final int SENSCORD_ANGULAR_VELOCITY_NOT_SUPPORTED = 0;

        /**
         * Unit : [deg/s].
         */
        public static final int SENSCORD_DEGREE_PER_SECOND = 1;

        /**
         * Unit : [rad/s].
         */
        public static final int SENSCORD_RADIAN_PER_SECOND = 2;
    }

    public static interface MagneticFieldUnit {
        /**
         * Not supported.
         */
        public static final int SENSCORD_MAGNETIC_FIELD_NOT_SUPPORTED = 0;

        /**
         * Unit : [gauss].
         */
        public static final int SENSCORD_GAUSS = 1;

        /**
         * Unit : [uT].
         */
        public static final int SENSCORD_MICRO_TESLA = 2;
    }

    public static interface OrientationUnit {
        /**
         * Not supported.
         */
        public static final int SENSCORD_ORIENTATION_NOT_SUPPORTED = 0;

        /**
         * Unit : [deg].
         */
        public static final int SENSCORD_DEGREE = 1;

        /**
         * Unit : [rad].
         */
        public static final int SENSCORD_RADIAN = 2;
    }


    /**
     * Get acceleration unit.
     * 
     * @return acceleration unit from {@link AccelerationUnit}
     */
    public int GetAccelerationUnit() {
        return acceleration;
    }

    /**
     * Get angular velocity unit.
     * 
     * @return angular velocity unit from {@link AngularVelocityUnit}
     */
    public int GetAngularVelocityUnit() {
        return angular_velocity;
    }

    /**
     * Get magnetic field unit.
     * 
     * @return magnetic field unit from {@link MagneticFieldUnit}
     */
    public int GetMagneticFieldUnit() {
        return magnetic_field;
    }

    /**
     * Get orientation unit.
     * 
     * @return orientation unit from {@link OrientationUnit}
     */
    public int GetOrientationUnit() {
        return orientation;
    }

    /**
     * Get acceleration unit String value.
     * 
     * @param acceleration acceleration value
     * @return acceleration String value
     */
    public String GetAccelerationUnitString(int acceleration) {
        String ret = "";

        switch (acceleration) {
            case AccelerationUnit.SENSCORD_ACCELERATION_NOT_SUPPORTED:
                ret = "NOT_SUPPORTED";
                break;
            case AccelerationUnit.SENSCORD_GRAVITATIONAL:
                ret = "GRAVITATIONAL";
                break;
            case AccelerationUnit.SENSCORD_METER_PER_SECOND_SQUARED:
                ret = "METER_PER_SECOND";
                break;
            default:
                break;
        }

        return ret;
    }

    /**
     * Get angular velocity unit String value.
     * 
     * @param angular_velocity angular velocity value
     * @return angular velocity String value
     */
    public String GetAngularVelocityUnitString(int angular_velocity) {
        String ret = "";

        switch (angular_velocity) {
            case AngularVelocityUnit.SENSCORD_ANGULAR_VELOCITY_NOT_SUPPORTED:
                ret = "NOT_SUPPORTED";
                break;
            case AngularVelocityUnit.SENSCORD_DEGREE_PER_SECOND:
                ret = "DEGREE_PER_SECOND";
                break;
            case AngularVelocityUnit.SENSCORD_RADIAN_PER_SECOND:
                ret = "RADIAN_PER_SECOND";
                break;
            default:
                break;
        }

        return ret;
    }

    /**
     * Get magnetic field unit String value.
     * 
     * @param magnetic_field magnetic field value
     * @return magnetic filed String value
     */
    public String GetMagneticFieldUnitString(int magnetic_field) {
        String ret = "";

        switch (magnetic_field) {
            case MagneticFieldUnit.SENSCORD_MAGNETIC_FIELD_NOT_SUPPORTED:
                ret = "NOT_SUPPORTED";
                break;
            case MagneticFieldUnit.SENSCORD_GAUSS:
                ret = "GAUSS";
                break;
            case MagneticFieldUnit.SENSCORD_MICRO_TESLA:
                ret = "TESLA";
                break;
            default:
                break;
        }

        return ret;
    }

    /**
     * Get orientation unit String value.
     * 
     * @param orientation orientation field value.
     * @return orientation String value
     */
    public String GetOrientationUnitString(int orientation) {
        String ret = "";

        switch (orientation) {
            case OrientationUnit.SENSCORD_ORIENTATION_NOT_SUPPORTED:
                ret = "NOT_SUPPORTED";
                break;
            case OrientationUnit.SENSCORD_DEGREE:
                ret = "DEGREE";
                break;
            case OrientationUnit.SENSCORD_RADIAN:
                ret = "RADIAN";
                break;
            default:
                break;
        }

        return ret;
    }

    /**
     * Set acceleration unit from {@link AccelerationUnit}.
     * 
     * @param acceleration acceleration value
     */
    public void SetAccelerationUnit(int acceleration) {
        this.acceleration = acceleration;
    }

    /**
     * Set angular velocity unit from {@link AngularVelocityUnit}.
     * 
     * @param angular_velocity angular velocity value
     */
    public void SetAngularVelocityUnit(int angular_velocity) {
        this.angular_velocity = angular_velocity;
    }

    /**
     * Set magnetic field unit from {@link MagneticFieldUnit}.
     * 
     * @param magnetic_field magnetic field value
     */
    public void SetMagneticFieldUnit(int magnetic_field) {
        this.magnetic_field = magnetic_field;
    }

    /**
     * Set orientation unit from {@link OrientationUnit}.
     * 
     * @param orientation orientation value
     */
    public void SetOrientationUnit(int orientation) {
        this.orientation = orientation;
    }
}
